keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,858 @@
+
+The `http` field has type `HTTPDownscalePodPicker` with the following specification:
+
+| Field         | Type             | Description                                                                                    |
+|---------------|------------------|------------------------------------------------------------------------------------------------|
+| `host`        | string           | Host name to connect to.                                                                       |
+| `httpHeaders` | HTTPHeader array | Custom headers to set in the request.                                                          |
+| `path`        | string           | Path to access on the HTTP server.                                                             |
+| `port`        | string / int     | Name or number of the port to access on the container. Number must be in the range 1 to 65535. |
+| `scheme`      | string           | Scheme to use for connecting to the host: `HTTP`, `HTTPS`. (default: `"HTTP"`)                 |
+
+The `httpHeaders` field has type [`HTTPHeader array`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#httpheader-v1-core)
+with the following specification:
+
+(NOTE: We will need to evaluate how feasible it is to extend `HTTPHeader v1 core` with `valueFrom` to enable mounting headers from Secrets,
+for authentication. This may need to be its own independent KEP.)
+
+| Field       | Type             | Description                                                            |
+|-------------|------------------|------------------------------------------------------------------------|
+| `name`      | string           | The header field name                                                  |
+| `value`     | string           | The header field value                                                 |
+| `valueFrom` | HTTPHeaderSource | Source for the header's value. Cannot be used if `value` is non-empty. |
+
+The `valueFrom` field has type `HTTPHeaderSource` with the following specification:
+
+| Field          | Type              | Description                                             |
+|----------------|-------------------|---------------------------------------------------------|
+| `secretKeyRef` | SecretKeySelector | Selects a key of a secret in the ReplicaSet's namespace |
+
+The `secretKeyRef` field has type [`SecretKeySelector`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#secretkeyselector-v1-core):
+
+| Field  | Type   | Description                                                       |
+|--------|--------|-------------------------------------------------------------------|
+| `key`  | string | The key of the secret to select from. Must be a valid secret key. |
+| `name` | string | Name of the referent.                                             |
+| `key`  | string | Specify whether the Secret or its key must be defined.            |
+
+
+#### "Pod Picker" API contract:
```

Comment:
This should be versioned

Reply:
100% agree, and I think we can call the first API version `v1alpha1`, while the feature is in alpha.

(For example, when implementing this contract in HTTP, we would wrap these payloads into JSON objects, which have an "api_version" field at the root, in case we change the contract in the future.)

Reply:
Where is the version specified?  In the type, in the URL path, somewhere else?

Reply:
@thockin after thinking about this more, I am not sure we need to implement versions at this stage.

This is because we can make it so that we never make breaking changes to this contract:
- REQUEST payloads sent by kube-controller:
   - Pod-Picker implementations can safely assume we will always send `number_of_pods_requested` and `candidate_pods`
   - if in the future we add other fields (for example, that give more pod information), Pod-Picker implementations can just ignore those fields, and use `number_of_pods_requested` and `candidate_pods`
- RESPONSE payloads sent by the Pod-Picker API:
   - while I can't foresee a situation where `chosen_pods` and `tied_pods` are not sufficient, if we ever need to change the response structure, we could implement versioning at THAT time, and have "unversioned" responses assumed as version 1